lib/index.js

```diff
@@ -120,163 +220,205 @@ export default {
-          linterResults.push({
-            severity: mapSeverity[issue.severity] || 'warning',
-            excerpt: `${issue.engine_name.toUpperCase()}: ${issue.description} [${issue.check_name}]`,
-            description: (issue.content && issue.content.body) ? issue.content.body : undefined,
-            location: {
-              file: filePath,
-              position: range,
-            },
-          });
-        });
-
-        // Log the length of time it took to run analysis
-        endMeasure(measureId);
-        return linterResults;
-      },
+      lint: async textEditor => this.debouncedLint(textEditor),
+    };
+  },
+
+  /**
+   * @summary Debounces the linting to join triggerings from multiple files of same project.
+   * @param   {TextEditor} textEditor The TextEditor instance of the triggering file.
+   * @return  {Promise}               An array of issues in the format that AtomLinter expects.
+   */
+  async debouncedLint(textEditor) {
+    const path = textEditor.getPath();
+
+    // Exit early on `untitled` files (not saved into disk yet)
+    if (path === undefined) return null;
+
+    if (!this.cache[path]) {
+      // Beware with race condition: textEditor observer and linter fired simultaneously
+      await this.cacheEditor(textEditor);
+    }
+    const { project } = this.cache[path];
+    const now = Date.now();
+    if (this.linting[project] === undefined) {
+      this.linting[project] = [now];
+    } else {
+      this.linting[project].push(now);
+    }
+
+    await delay(250);
+    this.linting[project].shift();
+
+    // More lints for the same project have been requested and delayed.
+    if (this.linting[project].length > 0) return null;
+
+    // This is the last requested lint, so analyze!
+    return this.lintProject(project);
+  },
+
+  /**
+   * @summary Lints a project.
+   * @param   {String}  path The absolute path to the project to analyze.
+   * @return  {Promise}      An array of issues in the format that AtomLinter expects.
+   */
+  async lintProject(path) {
+    // Debug the command executed to run the Code Climate CLI to the console
+    devLog(`Analyzing project @ ${path}`);
+
+    // Start measure for how long the analysis took
+    measure.start(path);
+
+    // Exec cc-cli and handle unique spawning (killed execs will return `null`)
+    const result = await this.runCli(path);
+    if (result === null) return null;
+
+    const linterResults = this.parseIssues(path, result);
+
+    // Log the length of time it took to run analysis
+    measure.end(path);
+
+    this.reset(path);
+    return linterResults;
+  },
+
+  /**
+   * @summary Cache and keeps track of open textEditors and cache its file/project paths.
+   * @param   {TextEditor} textEditor TextEditor instance of the file which triggered the analysis.
+   */
+  async cacheEditor(textEditor) {
+    const path = textEditor.getPath();
+
+    if (this.cache[path]) return;
+
+    textEditor.onDidDestroy(() => delete this.cache[path]);
```

Comment:
Nice solution to keeping the cache but invalidating it when needed. The only potential issue is that you need to remember to re-populate it if the `TextEditor`'s path is missing from the cache (which you did where needed 👍).